consumer/consumererror/error.go

```diff
@@ -0,0 +1,136 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package consumererror // import "go.opentelemetry.io/collector/consumer/consumererror"
+
+import (
+	"errors"
+	"net/http"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.opentelemetry.io/collector/consumer/consumererror/internal/statusconversion"
+)
+
+// Error is intended to be used to encapsulate various information that can add
+// context to an error that occurred within a pipeline component. Error objects
+// are constructed through calling `New` with the relevant options to capture
+// data around the error that occurred.
+//
+// Error should be obtained from a given `error` object using `errors.As`.
+type Error struct {
+	error
+	httpStatus  int
+	grpcStatus  *status.Status
+	isRetryable bool
+}
+
+var _ error = (*Error)(nil)
+
+// NewOTLPHTTPError records an HTTP status code that was received from a server
+// during data submission.
+func NewOTLPHTTPError(origErr error, httpStatus int) error {
+	var retryable bool
+	switch httpStatus {
+	case http.StatusTooManyRequests, http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout:
+		retryable = true
+	}
+
+	return &Error{error: origErr, httpStatus: httpStatus, isRetryable: retryable}
+}
+
+// NewOTLPGRPCError records a gRPC status code that was received from a server
+// during data submission.
+func NewOTLPGRPCError(origErr error, status *status.Status) error {
+	var retryable bool
+	if status != nil {
+		switch status.Code() {
+		case codes.Canceled, codes.DeadlineExceeded, codes.Aborted, codes.OutOfRange, codes.Unavailable, codes.DataLoss:
+			retryable = true
+		}
+	}
+
+	return &Error{error: origErr, grpcStatus: status, isRetryable: retryable}
+}
+
+// NewRetryableError records that this error is retryable according to OTLP specification.
+func NewRetryableError(origErr error) error {
+	return &Error{error: origErr, isRetryable: true}
+}
+
+// Error implements the error interface.
+func (e *Error) Error() string {
+	return e.error.Error()
```

Comment:
What happens when e.error is nil? I imagine we should see codes.OK and HTTP Status 200 in this case.
I couldn't remember details, but I went looking for `codes.OK` appearing in collector-contrib and it's always been a bit confusing to me, knowing when to use a Status with codes.OK vs a nil error. the OTLP exporter has code like

```
	// We have an error, check gRPC status code.
	st := status.Convert(err)
	if st.Code() == codes.OK {
		// Not really an error, still success.
		return nil
	}
```

Are callers expected to check this themselves, or can we bake it into this library?

Reply:
Missed this, and asked a very similar question.